src/primitives.rs

```diff
@@ -0,0 +1,161 @@
+use super::requests;
+
+use crate::requests::RequestApiVersions;
+use bytes::Buf;
+use core::borrow::{Borrow, BorrowMut};
+use std::fmt::{Error, Formatter};
+use DecodeErrors::BadData;
+
+enum ApiKeys {
+    Metadata = 3,
+    ApiVersions = 18,
+}
+
+pub enum ApiVersions {
+    Version0,
+    Version1,
+    Version2,
+}
+
+#[derive(Debug)]
+pub enum DecodeErrors {
+    BufferUnderflow,
+    BufferOverflow,
+    BadData,
+}
+
+pub trait Message {
+    fn get_size(&self) -> i32;
+}
+
+pub trait FromByte: Default {
+    fn decode(&mut self, buf: &mut Buf) -> Result<(), DecodeErrors>;
+
+    fn decode_new(buf: &mut Buf) -> Result<Self, DecodeErrors> {
+        let mut temp = Self::default();
+        match temp.decode(buf) {
+            Ok(_) => Ok(temp),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+pub trait DecodableMessage: Message + FromByte {}
+
+pub trait Request: DecodableMessage {
+    fn get_header(&self) -> &HeaderRequest;
+}
+
+// Skip for now
+pub trait Response: DecodableMessage {}
+
+impl FromByte for i8 {
+    fn decode(&mut self, buf: &mut Buf) -> Result<(), DecodeErrors> {
+        *self = buf.get_i8();
+        Ok(())
+    }
+}
+
+impl FromByte for i16 {
+    fn decode(&mut self, buf: &mut Buf) -> Result<(), DecodeErrors> {
+        *self = buf.get_i16_be();
+        Ok(())
+    }
+}
+
+impl FromByte for i32 {
+    fn decode(&mut self, buf: &mut Buf) -> Result<(), DecodeErrors> {
+        *self = buf.get_i32_be();
+        Ok(())
+    }
+}
+
+impl FromByte for String {
+    fn decode(&mut self, buf: &mut Buf) -> Result<(), DecodeErrors> {
+        let mut length: i16 = 0;
+        length.decode(buf);
+        if length == 0 || length == -1 {
+            return Ok(());
+        }
+        // Length should not be smaller than -1 according to the Kafka protocol
+        if length < -1 {
+            // ToDo: throw error
+            return Ok(());
+        }
+        self.reserve(length as usize);
+        // Any smart way to read a String from buffer?
+        let mut temp_vec: Vec<u8> = vec![0; length as usize];
+        buf.copy_to_slice(temp_vec.as_mut_slice());
+        // ToDo: throw error
+        *self = String::from_utf8_lossy(temp_vec.as_ref()).parse().unwrap();
+
+        if self.len() != length as usize {
+            // ToDo: throw error
+            return Ok(());
+        }
+        Ok(())
+    }
+}
+
+// Reference: https://github.com/spicavigo/kafka-rust/blob/c58cf5f30b35fad6ab163416d51d2b99a30da9c2/src/protocol/mod.rs#L108
+#[derive(Debug)]
+pub struct HeaderRequest {
+    pub api_key: i16,
```

Comment:
This probably isn't a major concern right now, but eventually you'll want to think about your library's API surface a bit. It's worth noting that having all the fields on a struct be `pub` means that anyone using your library can construct new instances of your struct. Instead, it's probably best for the fields to be private, and provide methods for accessing them. That way, all valid instances of the type will come from the library.